packages/loader/container-loader/src/loader.ts

```diff
@@ -468,17 +482,40 @@ export class Loader implements IHostLoader {
 			container = await this.loadContainer(request, resolvedAsFluid, pendingLocalState);
 		}
 
-		if (container.deltaManager.lastSequenceNumber <= fromSequenceNumber) {
+		if (shouldFreeze) {
+			if (
+				opsBeforeReturn === "sequenceNumber" &&
+				container.deltaManager.lastSequenceNumber > fromSequenceNumber
+			) {
+				// If we are trying to freeze the container at a specific sequence number, then we need to throw an
+				// error if the latest snapshot is past the specified sequence number.
+				throw new Error(
+					"Cannot satisfy request to freeze the container at the specified sequence number. Most recent snapshot is newer than the specified sequence number.",
+				);
+			}
+			// Force readonly mode - this will ensure we don't receive an error for the lack of join op
+			container.forceReadonly(true);
+		}
+
+		// If we have not yet reached `fromSequenceNumber`, we will wait for ops to arrive until we reach it
+		if (container.deltaManager.lastSequenceNumber < fromSequenceNumber) {
 			await new Promise<void>((resolve, reject) => {
 				function opHandler(message: ISequencedDocumentMessage) {
-					if (message.sequenceNumber > fromSequenceNumber) {
+					if (message.sequenceNumber >= fromSequenceNumber) {
+						if (shouldFreeze) {
+							// Disconnect the container now that we have finished loading.
+							container.disconnect();
```

Comment:
The logic is all over the place. If we do something (like disconnects), can we do it in one place? I see similar code in container.ts, which feels like it points to - all this logic should be in one layer (not spread out between Loader & Container). We also intend to deprecate Loader, so it's another signal that it should be outside of Loader code (preferred) or in Container.

Reply:
Should be doable here - one of the other comments talks about removing the disconnect (instead, only pause) which will reduce the logic, and I think everything can probably live inside the "freezeAtSeqNumListener" in container.ts.

Preparing for Loader deprecation is mostly about making sure we're not adding any new smarts to Loader - it's already almost nothing today, once caching is removed it will be pretty much irrelevant and we can direct people to instantiate containers more directly, maybe pending some thinking on the container surface area..